pkg/segment/writer/segwriter.go

```diff
@@ -1042,29 +1044,76 @@ func BackFillPQSSegmetaEntry(segsetkey string, newpqid string) {
 			continue
 		}
 	}
-	wfd, err := os.OpenFile(localSegmetaFname, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	err = WriteSegMeta(localSegmetaFname, preservedSmEntries)
 	if err != nil {
-		log.Errorf("BackFillPQSSegmetaEntry: Failed to open SegMetaFile name=%v, err:%v", localSegmetaFname, err)
-		return
+		log.Errorf("BackFillPQSSegmetaEntry: failed to write Segmeta: err=%v", err)
+	}
+}
+
+func WriteSegMeta(segmetaFname string, segMetaEntries []*structs.SegMeta) error {
+	wfd, err := os.OpenFile(segmetaFname, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("WriteSegMeta: Failed to open SegMetaFile name=%v, err:%v", segmetaFname, err)
 	}
 	defer wfd.Close()
 
-	for _, smentry := range preservedSmEntries {
+	for _, smentry := range segMetaEntries {
 
 		segmetajson, err := json.Marshal(*smentry)
 		if err != nil {
-			log.Errorf("BackFillPQSSegmetaEntry: failed to Marshal: err=%v", err)
-			return
+			return fmt.Errorf("WriteSegMeta: failed to Marshal: segmeta filename=%v: err=%v", segmetaFname, err)
 		}
```

Comment:
I'm not sure this `err` will give us enough info to debug. I think you should include in this log message the `smentry` that you tried to marshal

Also I'd recommend both logging and returning the error, just to make sure that we do log this error even if the caller of this function doesn't log the error. You can do this with:
```
return utils.TeeErrorf("WriteSegMeta: failed to Marshal: segmeta filename=%v: err=%v", segmetaFname, err)
```

Reply:
Actually, the callers of this method are logging the error so that we know what was the source and prevent duplicate logs. What do you think?

Reply:
addressed in the latest commit.